### PR TITLE
Implement Inventory Download Lambda

### PR DESCRIPTION
**Pull request by simonkrol**
*Issue #, if available: [Asana](https://app.asana.com/0/1202938635037719/1204098877654449/f)

*Description of changes:*
- Implements the lambda which will download the inventory byte ranges from Glacier and upload it to S3
- Updates the typing of responses from the previously mentioned lambda handlers.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: source/refreezer/application/glacier_s3_transfer/upload.py [hunk 1]
```diff
@@ -69,7 +68,9 @@ def complete_upload(self) -> CompleteMultipartUploadOutputTypeDef:
         )
 
     @staticmethod
-    def _build_part(part_number: int, etag: str, checksum: str) -> CompletedPartTypeDef:
+    def _build_part(
```
Comment (kamyarz-aws) [inline source/refreezer/application/glacier_s3_transfer/upload.py:71]: If it is staticmethod it should be defined as private (with _ ) function

File: source/refreezer/application/glacier_s3_transfer/facilitator.py [hunk 2]
```diff
@@ -45,7 +47,7 @@ def __init__(
 
         self.ignore_glacier_checksum = ignore_glacier_checksum
 
-    def transfer(self) -> CompletedPartTypeDef:
+    def transfer(self) -> responses.GlacierRetrieval:
```
Comment (kamyarz-aws) [inline source/refreezer/application/glacier_s3_transfer/facilitator.py:50]: transfer is not descriptive enough name to me. Either change the function name  or add docstrings (preferably this one ) for better documentation please.

File: source/refreezer/application/handlers.py [hunk 3]
```diff
@@ -79,12 +78,25 @@ def inventory_chunk_lambda_handler(
...
+        event["VaultName"],
+        event["StartByte"],
+        event["EndByte"],
+        event["VaultName"],
+        event["S3DestinationBucket"],
+        event["S3DestinationKey"],
+        event["UploadId"],
+        event["PartNumber"],
+        True,
+    )
```
Comment (kamyarz-aws) [inline source/refreezer/application/handlers.py:97]: `
a_list = []
for key in ["JobId", "VaultName", "StartByte", "EndByte", "VaultName", "S3DestinationBucket", "S3DestinationKey", "UploadId", "PartNumber"]:
     a_list.append(event[key])
a_list.append(True)
GlacierToS3Facilitator(*a_list)
`
Might end up with less code line duplication and and fewer lines of code. You could give it a try.

File: source/refreezer/infrastructure/stack.py [hunk 4]
```diff
@@ -597,7 +597,10 @@ def __init__(
                     "id": "AwsSolutions-IAM5",
                     "reason": "By default wildcard permission is granted to the state machine from various sources.  This will be replaced with a proper IAM role later.",
                     "appliesTo": [
-                        f"Resource::<{inventory_chunk_download_lambda_function_logical_id}.Arn>:*"
+                        f"Resource::<{inventory_download_lambda_function_logical_id}.Arn>:*",
+                        "Action::*",
+                        "Action::s3:Abort*",
+                        "Resource::<InventoryBucketA869B8CB.Arn>/*",
```
[the remark this answers is not in the document]
Comment (simonkrol) [reply source/refreezer/infrastructure/stack.py:611]: I think you're right, will remove

File: source/refreezer/application/glacier_s3_transfer/upload.py (hunk 1 above)
Comment (simonkrol) [reply source/refreezer/application/glacier_s3_transfer/upload.py:71]: It is a static method, such that it doesn't depend on any class or instance values, but it also isn't expected to be used outside of this class.

**Review by bassemwanis — DISMISSED** [review]

**Review by bassemwanis — APPROVED** [review]

**Pull request merged**